jmx/jmx.go

```diff
@@ -39,38 +39,112 @@ const (
+	hostname           string
+	port               string
+	username           string
+	password           string
+	keyStore           string
+	keyStorePassword   string
+	trustStore         string
+	trustStorePassword string
+	remote             bool
+}
+
+func (cfg *connectionConfig) isSSL() bool {
+	return cfg.keyStore != "" && cfg.keyStorePassword != "" && cfg.trustStore != "" && cfg.trustStorePassword != ""
+}
+
+func (cfg *connectionConfig) command() []string {
+	c := make([]string, 0)
 	if os.Getenv("NR_JMX_TOOL") != "" {
 		c = strings.Split(os.Getenv("NR_JMX_TOOL"), " ")
 	} else {
 		c = []string{jmxCommand}
 	}
 
-	c = append(c, "--hostname", hostname, "--port", port)
-	if username != "" && password != "" {
-		c = append(c, "--username", username, "--password", password)
+	c = append(c, "--hostname", cfg.hostname, "--port", cfg.port)
+	if cfg.username != "" && cfg.password != "" {
+		c = append(c, "--username", cfg.username, "--password", cfg.password)
 	}
-
-	return
-}
-
-func getCommandWithSSL(hostname, port, username, password, keyStore, keyStorePassword, trustStore, trustStorePassword string) []string {
-	c := getCommand(hostname, port, username, password)
-
-	if keyStore != "" && keyStorePassword != "" && trustStore != "" && trustStorePassword != "" {
-		c = append(c, "--keyStore", keyStore, "--keyStorePassword", keyStorePassword, "--trustStore", trustStore, "--trustStorePassword", trustStorePassword)
+	if cfg.remote {
+		c = append(c, "--remote")
+	}
+	if cfg.isSSL() {
+		c = append(c, "--keyStore", cfg.keyStore, "--keyStorePassword", cfg.keyStorePassword, "--trustStore", cfg.trustStore, "--trustStorePassword", cfg.trustStorePassword)
 	}
 
 	return c
 }
 
 // Open will start the nrjmx command with the provided connection parameters.
+// Deprecated use OpenWithParameters instead.
 func Open(hostname, port, username, password string) error {
-	return OpenWithSSL(hostname, port, username, password, "", "", "", "")
+	config := &connectionConfig{
+		hostname: hostname,
+		port:     port,
+		username: username,
+		password: password,
+	}
+
+	return openConnection(config)
 }
 
-// OpenWithSSL will start the nrjmx command with the provided SSL connection parameters
+// OpenWithSSL will start the nrjmx command with the provided SSL connection parameters.
+// Deprecated use OpenWithParameters instead.
 func OpenWithSSL(hostname, port, username, password, keyStore, keyStorePassword, trustStore, trustStorePassword string) error {
+	config := &connectionConfig{
+		hostname:           hostname,
+		port:               port,
+		username:           username,
+		password:           password,
+		keyStore:           keyStore,
+		keyStorePassword:   keyStorePassword,
+		trustStore:         trustStore,
+		trustStorePassword: trustStorePassword,
+	}
+
+	return openConnection(config)
+}
+
+// OpenWithParameters executes a nrjmx command using the given options.
+func OpenWithParameters(hostname, port, username, password string, opts ...Option) error {
```

Comment:
either 1 approach or the other, why keeping so many construction option?
if you want opt funcs then remove OpenWithSSL

Reply:
This would be a breaking change for all the integrations using this version. The new major SDK version will need to remove `OpenWithSSL` and `Open` methods. This is why I commented them as deprecated. We should always use this option `OpenWithParameters`